src/pages/timecard/Timecard.jsx

```diff
@@ -6,20 +6,64 @@ import BackLink from '../../components/common/form/navigation/backlink/BackLink'
 import SelectTimecardPeriodType from '../../components/timecard/select-timecard-period-type/SelectTimecardPeriodType';
 import ErrorSummary from '../../components/common/form/error-summary/ErrorSummary';
 import generateDocumentTitle from '../../utils/generate-document-title/generateDocumentTitle';
+import {
+  getTimeEntries,
+  getTimePeriodTypeById,
+} from '../../api/services/timecardService';
+import {
+  formatTime,
+  formatDate,
+} from '../../utils/time-entry-utils/timeEntryUtils';
+import { UrlSearchParamBuilder } from '../../utils/api-utils/UrlSearchParamBuilder';
 import EditShiftTimecard from '../../components/timecard/edit-shift-timecard/EditShiftTimecard';
 import { useTimecardContext } from '../../context/TimecardContext';
 
 import { sortErrorKeys } from '../../utils/sort-errors/sortErrors';
 
+const updateTimeEntryContextData = async (setTimeEntries) => {
+  const timeEntriesParams = new UrlSearchParamBuilder()
+    .setTenantId('00000000-0000-0000-0000-000000000000')
+    .setFilter('ownerId==1')
+    .getUrlSearchParams();
+  const timeEntriesResponse = await getTimeEntries(timeEntriesParams);
+
+  if (timeEntriesResponse.data.items?.length > 0) {
+    const existingTimeEntries = await Promise.all(
+      timeEntriesResponse.data.items.map(async (timeEntry) => {
+        const timePeriodTypeParams = new UrlSearchParamBuilder()
+          .setTenantId('00000000-0000-0000-0000-000000000000')
+          .getUrlSearchParams();
+        const timePeriodType = await getTimePeriodTypeById(
```

Comment:
I think I’m right in saying that this will do a fresh API call for each TimeEntry? I guess this will require a lot of duplicate API calls?

My feeling is that we should aim to minimise the number of API calls our UI has to do where possible, as these will have an impact on the performance of our app and the load on our database.

TimePeriodTypes are likely to change very rarely so I’m thinking we want to cache these responses in some way. _Where_ we cache it is an interesting question. I think in the case of TimePeriodTypes, which will be consistent for a wide range of users, there’s an argument for caching at API level so all users can benefit from the same cache, but that’ll want some thought and is perhaps an optimisation that can come later.

Potentially we get a good effort/value trade-off if we do a bit of caching within the UI, perhaps with a map of cached responses within our function in timecard service?

Interested to get your views and happy to discuss this. Possibly even worth a tech huddle chat 😄

Reply:
For the record, we decided to retrieve the TimePeriodTypes just once on app load and make it available via an ‘ApplicationContext’